### ISSUE TITLE
dpcpp fails to resolve `enable_if<std::numeric_limits<T>::digits == 64>`

### ISSUE THREAD

**Issue by ndevenish** [intel/llvm#6488]
**Describe the bug**
dpcpp fails to correctly resolve TMP struct condition instantiation.

**To Reproduce**
With `test.cc`:
```c++
#include <iostream>
#include <limits>

template <typename T, typename Enable = void> struct float_info;

template <typename T>
struct float_info<T, typename std::enable_if<std::numeric_limits<T>::digits == 64>::type> {
    static const int can_access = 1;
};

int main(int argc, char **argv) {
    constexpr bool check = std::numeric_limits<long double>::digits == 64;
    std::cout << "Constexpr: " << check << std::endl;
    std::cout << "Digits:    " << std::numeric_limits<long double>::digits << std::endl;

    std::cout << float_info<long double>::can_access << std::endl;
    return 0;
}
```
And `Dockerfile`:
```
FROM intel/oneapi-basekit:2022.2-devel-ubuntu20.04

COPY test.cc /opt/
RUN g++   -std=c++17 /opt/test.cc
RUN dpcpp -std=c++17 /opt/test.cc
```
Gets the error (from dpcpp, g++ within the base image builds the sources without issue):
```
/opt/test.cc:16:18: error: implicit instantiation of undefined template 'float_info<long double>'
    std::cout << float_info<long double>::can_access << std::endl;
                 ^
/opt/test.cc:4:54: note: template is declared here
template <typename T, typename Enable = void> struct float_info;
                                                     ^
1 error generated.
```
Removing the line that tries to reference the field (`float_info<long double>::can_access`) looks as though the conditions are checking the correct lookup:
```
RUN ./a.out
Constexpr: 1
Digits:    64
```

This same example also works fine with a selection of compilers on [godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:18,positionColumn:2,positionLineNumber:18,selectionStartColumn:2,selectionStartLineNumber:18,startColumn:2,startLineNumber:18),source:'%23include+%3Ciostream%3E%0A%23include+%3Climits%3E%0A%0Atemplate+%3Ctypename+T,+typename+Enable+%3D+void%3E+struct+float_info%3B%0A%0Atemplate+%3Ctypename+T%3E%0Astruct+float_info%3CT,+typename+std::enable_if%3Cstd::numeric_limits%3CT%3E::digits+%3D%3D+64%3E::type%3E+%7B%0A++++static+const+int+can_access+%3D+1%3B%0A%7D%3B%0A%0Aint+main(int+argc,+char+**argv)+%7B%0A++++constexpr+bool+check+%3D+std::numeric_limits%3Clong+double%3E::digits+%3D%3D+64%3B%0A++++std::cout+%3C%3C+%22Constexpr:+%22+%3C%3C+check+%3C%3C+std::endl%3B%0A++++std::cout+%3C%3C+%22Digits:++++%22+%3C%3C+std::numeric_limits%3Clong+double%3E::digits+%3C%3C+std::endl%3B%0A%0A++++std::cout+%3C%3C+float_info%3Clong+double%3E::can_access+%3C%3C+std::endl%3B%0A++++return+0%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:clang1400,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B20',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+clang+14.0.0+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),k:27.932351336606654,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+14.0.0+(Compiler+%231)',t:'0')),k:38.73431533006,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Environment (please complete the following information):**

- OS: RH7 with [GCC 7.4](https://www.intel.com/content/www/us/en/developer/articles/technical/oneapi-dpcpp-fpga-designs-on-rhel-7.html), but also the latest `intel/oneapi-basekit` docker image.
- DPC++ version: `Intel(R) oneAPI DPC++/C++ Compiler 2022.1.0 (2022.1.0.20220316)` 
- Dependencies version: [e.g. low-level runtime versions (like NEO 20.04)]

**Additional context**
Traced down while trying to work out a library error: https://github.com/fmtlib/fmt/issues/3005. This is not the first time I've encountered apparent dpcpp problems involving this library ([intel forums "dpcpp gives "InvalidModule: Invalid SPIR-V..." on linking"](https://community.intel.com/t5/Intel-oneAPI-Data-Parallel-C/dpcpp-gives-quot-InvalidModule-Invalid-SPIR-V-quot-on-linking/m-p/1313290)) so it might be worth considering adding to the test suite.

**Comment by bader** [comment]
@ndevenish, `long double` is not supported by the target.

**Comment by ndevenish** [comment]
I'm don't think that appears to be the problem here? You appear to be able to declare `long double`, get it's size, do maths with it, query it's properties via numerical_limits:

```c++
    std::cout << "Digits:    " << std::numeric_limits<long double>::digits << std::endl;
    std::cout << "Sizes: " << sizeof(double) << " vs " << sizeof(long double) << std::endl;
    long double a = 33.4;
    std::cout << "Using it: " << a + 4 << std::endl;
```
Giving
```
Digits:    64
Sizes: 8 vs 16
Using it: 37.4
```
In fact, you can manually declare the correct specialisation:
```c++
template<> struct float_info<long double> {
    static const int can_access = 2;
};
```
and then the original case works fine. It's also mentioned on the [dpcpp Reference Guide](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-dpcpp-cpp-compiler-dev-guide-and-reference/top/compiler-reference/floating-point-operations/handle-floating-point-array-operations-loop-body.html) as one of the possible floating point types?

**Comment by bader** [comment]
Right, but `long double` size for the default target (i.e. SPIR) is 128 bits and AFAIK, SPIR-V with 128 bit floating point data types are not supported by Intel's execution environments (i.e. Level Zero and OpenCL).
Because `sizeof(long double) == 128` for SPIR target, the error message is expected.

**Comment by ndevenish** [comment]
But compiling this code _literally_ prints `sizeof(long double) == 64`. Besides, it's also not even checking the `enable_if` condition based on `sizeof`, it's checking based on `std::numeric_limits<long double>::digits`, which when looked up separately gives the expected result 64.

Are you saying that the "default target" (not x86_64 apparently) is evaluated using a completely different standard library inside template evaluation vs outside?

In fact, it's not even true that the sizeof() is evaluating differently:

```
template <typename T>
struct float_info<T, typename std::enable_if<sizeof(T) == 16>::type> {
    static const int can_access = 16;
};
```
Works perfectly, and matches `long double`. Adding both the non-matching and matching:
```
template <typename T>
  struct float_info<T, typename std::enable_if<std::numeric_limits<T>::digits == 64>::type> {
  static const int can_access = 1;
};
template <typename T>
struct float_info<T, typename std::enable_if<sizeof(T) == 16>::type> {
    static const int can_access = 16;
};
```
Actually seems to resolve; it prints an `error: ambiguous template instantiation for ‘struct float_info<long double>’`.

**Comment by ndevenish** [comment]
I'm not trying to write any of this in a SPIR-V execution target, it's all in the host-side x86_64. Unless literally everything is always compiled to SPIR-V first, then translated, and somewhere in the middle it switched to a different stdlib lookup table for the purposes of resolving template enable conditions?

**Comment by bader** [comment]
I see that `std::numeric_limits<long double>::digits` for SPIR target is 53 if I use libstdc++. 

> I'm not trying to write any of this in a SPIR-V execution target, it's all in the host-side x86_64.

If so, I suggest using `icx` driver instead of `dpcpp` when there is no code to offload. I'll check if it's something we can fix for `dpcpp` driver.

`dpcpp` driver **parses** all sources for device targets to make sure the code is valid. The problem here is that "compile time" evaluations do not match host side. Even when it's expected, I would expect more clear diagnostics.

**Comment by ndevenish** [comment]
Okay - so, if I understand correctly, everything is essentially compiled twice, separately - for both targets - and the error is invisibly coming from the second target in this case? This would also explain the other error I encountered (["Invalid SPIR-V"](https://community.intel.com/t5/Intel-oneAPI-Data-Parallel-C/dpcpp-gives-quot-InvalidModule-Invalid-SPIR-V-quot-on-linking/m-p/1313290)) which was apparently fixed (though I don't have a commit/ticket reference for that).

Adding a case for `std::enable_if<std::numeric_limits<T>::digits == 53>` thus seems to solve the error because I guess satisfies the second target. I also guess two copies of the code are encoded into the binary - so that it "in principle" doesn't need to know in advance what is/isn't supposed to run on the GPU/accelerator and solves the problem (of identifying kernels) that way?

So, basically, the intention is that we are supposed to use icx for general host-only code, and dpcpp for the parts that actually deploy kernels? Does intel have any examples of this setup, there is almost zero mention of `icx` on the https://github.com/oneapi-src/oneAPI-samples and Ideally I'd be able to mix "real C++" host code with the parts that resolve to kernels....

**Comment by bader** [comment]
> Because `sizeof(long double) == 128` for SPIR target, the error message is expected.

I was wrong here. I think I confused the size enforced by OpenCL language which is 128 bits. I see that `sizeof(long double) == 8` i.e. 64 bits and it matches `sizeof(double)`. AFIAK, all GPU targets do the same and support only 64-bit representation of `long double` even if for "single source" programming models which might cause the difference in host and device representations. This issue impacts CUDA and HIP modes in clang as well. DPC++ emits errors if `long doubles` are being used in device code.

> Okay - so, if I understand correctly, everything is essentially compiled twice, separately - for both targets - and the error is invisibly coming from the second target in this case?

I specifically avoided "compiled" because compilation process consists of multiple stages and the process is different for the host and devices. For instance, the device compiler will not emit any 'device code' for the example we have in this issue as there is no code to offload, but before it knows that there is no code to offload, it has to parse the whole program assuming that it's a valid C++ code. Using undefined symbol is critical enough to break the compilation process before detecting that this program has no code to offload. You can learn more about how the compiler works in the documentation. https://github.com/intel/llvm/blob/sycl/sycl/doc/design/CompilerAndRuntimeDesign.md

> This would also explain the other error I encountered (["Invalid SPIR-V"](https://community.intel.com/t5/Intel-oneAPI-Data-Parallel-C/dpcpp-gives-quot-InvalidModule-Invalid-SPIR-V-quot-on-linking/m-p/1313290)) which was apparently fixed (though I don't have a commit/ticket reference for that).

To be honest, I don't see any connections between these two issues. Most likely someone else reported similar problem which was fixed independently.

> So, basically, the intention is that we are supposed to use icx for general host-only code, and dpcpp for the parts that actually deploy kernels? Does intel have any examples of this setup, there is almost zero mention of `icx` on the https://github.com/oneapi-src/oneAPI-samples and Ideally I'd be able to mix "real C++" host code with the parts that resolve to kernels....

No. What I proposed is the work-around for the issue with this particular example. The intention is that DPC++ can compile host-only code as well, but there are issues with compile-time computations relying on `long double` type traits to be the same for all "compilations". You can replace icx with any other C++ of your choice. DPC++ is required to compile SYCL code, but C++ code, which do not use SYCL can be compiled with other C++ compilers.

**Comment by ndevenish** [comment]
> To be honest, I don't see any connections between these two issues. Most likely someone else reported similar problem which was fixed independently.

"The SPIR-V target is different and has different rules, even if you are attempting to compile pure host code" seems to be exactly a common cause with the other issue. I'm also mildly troubled by the suggestion that despite two separate intel employees telling me the issue was identified, that it probably wasn't the case.

In any case, thank you for the documentation link and for helping me understand this; even if it is not a bug - and was an "invalid" question.
